packages/roosterjs-editor-plugins/lib/PasteManager/PasteManager.ts

```diff
@@ -59,129 +73,77 @@ export default class PasteManager implements EditorPlugin {
         // add undo snapshot before paste
         this.editor.addUndoSnapshot();
         let pasteEvent = <ClipboardEvent>(<PluginDomEvent>event).rawEvent;
-        let dataTransfer =
-            pasteEvent.clipboardData ||
-            (<WindowForIE>this.editor.getDocument().defaultView).clipboardData;
-        let clipboardData = getClipboardData(dataTransfer);
-        if ((clipboardData.textData || '') == '' && clipboardData.imageData.file) {
-            pasteEvent.preventDefault();
-            this.onPasteComplete(clipboardData);
+        buildClipBoardData(
+            pasteEvent,
+            this.editor,
+            data => this.paste(data, this.defaultPasteOption),
+            this.useDirectPaste ? removeUnsafeTags : null
+        );
+    }
+
+    paste = (clipboardData: ClipBoardData, pasteOption: PasteOption) => {
+        if (!clipboardData) {
+            return;
+        }
+
+        if (clipboardData.snapshotBeforePaste == null) {
+            clipboardData.snapshotBeforePaste = buildSnapshot(this.editor);
         } else {
-            // There is text to paste, so clear any image data if any.
-            // Otherwise both text and image will be pasted, this will cause duplicated paste
-            clipboardData.imageData = {};
-            this.retrieveHtml(this.editor, container => {
-                processImages(container, clipboardData);
-                clipboardData.htmlData = normalizeContent(container.innerHTML);
-
-                let document = this.editor.getDocument();
-                let fragment = document.createDocumentFragment();
-                let nodes = fromHtml(clipboardData.htmlData, document);
-                for (let node of nodes) {
-                    fragment.appendChild(node);
-                }
-
-                convertPastedContentFromWord(fragment);
-                this.editor.insertNode(fragment);
-                this.onPasteComplete(clipboardData);
-            });
+            restoreSnapshot(this.editor, clipboardData.snapshotBeforePaste);
```

Comment:
Not quite understand this: why build snapshot inside clipboardData and in what kind of cases we restore to snapshot before paste?

Reply:
This is for a new requirement. After pasted, user may want to change the way to paste, say we already pasted in HTML, but then user want to paste in plain text. So that we need to restore original content first, then paste again in plain text. Otherwise the content will be pasted twice.